### PR TITLE
tests: new store-state tool

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This is the new store tool including a new test. Also, it is included a network-state tool

A following pr will include the change in all the tests to start using the new tool

### CODE REVIEW

File: tests/lib/tools/store-state [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        fi
+        if ! command -v jq; then
+            #shellcheck source=tests/lib/core-config.sh
+            . "$TESTSLIB"/core-config.sh
+
+            SUFFIX="$(get_test_snap_suffix)"
+            snap install "jq$SUFFIX"
+        fi
+
+        # unsquash the snap to get it's name
```
Comment (MiguelPires) [inline tests/lib/tools/store-state:80]:
```suggestion
        # unsquash the snap to get its name
```

File: tests/lib/tools/store-state [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        unsquashfs -d /tmp/snap-squashfs "$snap_path" meta/snap.yaml
+        snap_name=$(yaml2json < /tmp/snap-squashfs/meta/snap.yaml | jq -r .name)
```
Comment (MiguelPires) [inline tests/lib/tools/store-state:82]: Should this be `local` as well?

File: tests/lib/tools/store-state [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+    # NOTE: snap id and snap name are derived from snap file name
+    SNAP_DECL_JSON_PARAM=""
+    SNAP_REV_JSON_PARAM=""
+    if [ -n "$snap_id" ]; then
+        SNAP_DECL_JSON_PARAM="--snap-decl-json=/tmp/snap-decl.json"
+        SNAP_REV_JSON_PARAM="--snap-rev-json=/tmp/snap-rev.json"
+    fi
+    p_decl=$(fakestore new-snap-declaration --dir "$dir" "$snap_path" "$SNAP_DECL_JSON_PARAM")
+    p_rev=$(fakestore new-snap-revision --dir "$dir" "$snap_path" "$SNAP_REV_JSON_PARAM")
```
Comment (MiguelPires) [inline tests/lib/tools/store-state:117]: There's a mix of lowercase and uppercase variables (which could be local as well I think). Is there a reason for that (like a convention or something)?

File: tests/lib/tools/store-state [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    mkdir -p "$top_dir/asserts"
+
+    # debugging
+    systemctl status fakestore || true
+    
+    # When a controlled store service is up
+    # Create fakestore at the given port
+    https_proxy=${https_proxy:-}
+    http_proxy=${http_proxy:-}
+    PORT="11028"
```
Comment (MiguelPires) [inline tests/lib/tools/store-state:147]: Same question

File: tests/lib/tools/network-state [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    done
+
+    # Ensure we really have the listen port, this will fail with an
+    # exit code if the port is not available.
+    ss -lnt | grep -Pq "LISTEN.*?:$PORT +.*?\\n*"
+}
+
+make_network_service() {
+    SERVICE_NAME="$1"
+    PORT="$2"
```
Comment (MiguelPires) [inline tests/lib/tools/network-state:30]: These could also be scoped

File: tests/lib/tools/network-state [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    echo "       network-state wait-listen-port <PORT>"
+    echo "       network-state make-network-service <SERVICE_NAME> <PORT>"
+}
+
+get_default_iface(){
+    ip route get 8.8.8.8 | awk '{ print $5; exit }'
+}
+
+wait_listen_port(){
+    PORT="$1"
```
Comment (MiguelPires) [inline tests/lib/tools/network-state:14]: And this as well

**Review by MiguelPires — COMMENTED** [review]
Looks good, I just had a few questions about local scoping and variable naming

File: tests/lib/tools/store-state (hunk 1 above)
Comment (sergiocazzolato) [reply tests/lib/tools/store-state:80]: done

File: tests/lib/tools/store-state (hunk 4 above)
Comment (sergiocazzolato) [reply tests/lib/tools/store-state:147]: fixed

File: tests/lib/tools/network-state (hunk 5 above)
Comment (sergiocazzolato) [reply tests/lib/tools/network-state:30]: fixed

File: tests/lib/tools/network-state (hunk 6 above)
Comment (sergiocazzolato) [reply tests/lib/tools/network-state:14]: fixed

File: tests/lib/external/snapd-testing-tools/remote/remote.wait-for [hunk 7]
```diff
@@ -8,8 +8,8 @@ DEFAULT_WAIT_FOR_SNAP_COMMAND_ATTEMPTS=200
 DEFAULT_WAIT_FOR_SNAP_COMMAND_WAIT=1
 DEFAULT_WAIT_FOR_DEV_INIT_ATTEMPTS=60
 DEFAULT_WAIT_FOR_DEV_INIT_WAIT=1
-DEFAULT_WAIT_FOR_REBOOT_ATTEMPTS=60
-DEFAULT_WAIT_FOR_REBOOT_WAIT=1
+DEFAULT_WAIT_FOR_REBOOT_ATTEMPTS=120
```
Comment (Meulengracht) [inline tests/lib/external/snapd-testing-tools/remote/remote.wait-for:11]: These values just feel incredibly arbitrary and I would love to see a comment explaining why these numbers were selected

File: tests/lib/tools/network-state [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    if [ -z "$(declare -f "$action")" ]; then
+        echo "network-state: no such command $subcommand" >&2
+        show_help
+        exit 1
+    fi
+
+    "$action" "$@"
+}
+
+main "$@"
```
Comment (Meulengracht) [inline tests/lib/tools/network-state:67]: missing newline at end of file

File: tests/lib/tools/store-state [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    if [ -z "$(declare -f "$action")" ]; then
+        echo "store-state: no such command $subcommand" >&2
+        show_help
+        exit 1
+    fi
+
+    "$action" "$@"
+}
+
+main "$@"
```
Comment (Meulengracht) [inline tests/lib/tools/store-state:216]: also newline missing at end of file

**Review by Meulengracht — APPROVED** [review]
+1, some minor comments

File: tests/lib/tools/network-state (hunk 8 above)
Comment (sergiocazzolato) [reply tests/lib/tools/network-state:67]: done

File: tests/lib/tools/store-state (hunk 9 above)
Comment (sergiocazzolato) [reply tests/lib/tools/store-state:216]: done

File: tests/lib/external/snapd-testing-tools/remote/remote.wait-for (hunk 7 above)
Comment (sergiocazzolato) [reply tests/lib/external/snapd-testing-tools/remote/remote.wait-for:11]: well, you are right, those numbers are arbitrary, they were selected trying to match with most of the wait times in the tests and also trying to follow common sense.

Comment (Meulengracht) [reply tests/lib/external/snapd-testing-tools/remote/remote.wait-for]: It's completely fine they are arbitrary, but just a comment saying what you just told me would be cool

**Pull request merged**